grype/db/internal/gormadapter/open.go

```diff
@@ -23,9 +25,18 @@ var readOptions = []string{
 
 // Open a new connection to a sqlite3 database file
 func Open(path string, write bool) (*gorm.DB, error) {
-	if write {
-		// the file may or may not exist, so we ignore the error explicitly
-		_ = os.Remove(path)
+	memory := len(path) == 0
+
+	if write && !memory {
+		if _, err := os.Stat(path); err == nil {
+			if err := os.Remove(path); err != nil {
```

Comment:
This really surprised me the first time I read it.

What I think is going on after thinking about it is: If we're opening the database for _write_ access, it's because we're making a whole new Grype DB and we should blow away the old one. But it requires a lot of knowledge of how Grype / Grype DB work to know that.

What if we split into two functions, `Create` and `Open`, and `Open` is always read-only? I think it would be less surprising for clients if calling `Create("/some/path")` when `/some/path` exists destroys the old data than calling `Open`.

I don't love that `""` as the path means, "using sqlite in memory," but I can't think of a good alternative right now.

Reply:
I didn't realize we were fitting into an existing interface. I can probably live with the current implementation in that case.

Reply:
On second thought, aren't we swapping out a bunch of the management of the database in the v6 change? In other words, this meets the interface we used for v5, but is it time to change that interface?

Reply:
Notes from sync review: we'll split this into two functions, `Create` which truncates the existing, and `Open` (or another name) that opens the database for read operations. Grype never updates a database, so Create and OpenForReading are the only logical operations.

Reply:
I came to a slightly different conclusion when thinking about this further: this function returns `*gorm.DB` which has read and write capabilities. Having separate readonly and write functions doesn't really help here if the gorm type can do both reads and writes anyway. We really want a helper to make clear what the connection string will be and add clear options (replace `write` with `truncate`).

I made a config object to encapsulate and test this logic which is controlled by functional options (currently only one).